src/main/java/org/folio/services/financedata/FinanceDataService.java

```diff
@@ -1,39 +1,191 @@
+      log.info("Finance data collection is empty, nothing to update");
+      return succeededFuture();
+    }
+
+    validateFinanceDataCollection(financeDataCollection);
+    return processAllocationTransaction(financeDataCollection, requestContext)
+      .compose(v -> updateFinanceData(financeDataCollection, requestContext))
+      .onSuccess(asyncResult -> processLogs(financeDataCollection, requestContext, COMPLETED))
+      .onFailure(asyncResult -> processLogs(financeDataCollection, requestContext, ERROR));
+  }
+
+  private void validateFinanceDataCollection(FyFinanceDataCollection financeDataCollection) {
+    for (int i = 0; i < financeDataCollection.getFyFinanceData().size(); i++) {
+      var financeData = financeDataCollection.getFyFinanceData().get(i);
+      validateFinanceDataFields(financeData, i);
+      var allocationChange = BigDecimal.valueOf(financeData.getBudgetAllocationChange());
+      var initialAllocation = BigDecimal.valueOf(financeData.getBudgetInitialAllocation());
+
+      if (allocationChange.doubleValue() < 0 && allocationChange.abs().doubleValue() > initialAllocation.doubleValue()) {
+        var params = List.of(new Parameter().withKey(String.format("financeData[%s].budgetAllocationChange", i))
+          .withValue(String.valueOf(financeData.getBudgetAllocationChange())));
+        var error = new Error().withMessage("Allocation change cannot be greater than initial allocation").withParameters(params);
+        throw new HttpException(422, new Errors().withErrors(List.of(error)));
+      }
+    }
+  }
+
+  private Future<Void> processAllocationTransaction(FyFinanceDataCollection fyFinanceDataCollection,
+                                                    RequestContext requestContext) {
+    var transactionsFuture = fyFinanceDataCollection.getFyFinanceData().stream()
+      .map(financeData -> createAllocationTransaction(financeData, requestContext))
+      .toList();
+
+    return GenericCompositeFuture.join(transactionsFuture)
+      .compose(compositeFuture -> {
+        List<Transaction> transactions = compositeFuture.list();
+        return createBatchTransaction(transactions, requestContext);
+      });
+  }
+
+  private Future<Transaction> createAllocationTransaction(FyFinanceData financeData, RequestContext requestContext) {
+    var allocation = calculateAllocation(financeData);
+    var transaction = new Transaction()
+      .withTransactionType(Transaction.TransactionType.ALLOCATION)
+      .withId(UUID.randomUUID().toString())
+      .withAmount(allocation)
+      .withFiscalYearId(financeData.getFiscalYearId())
+      .withToFundId(financeData.getFundId())
+      .withSource(Transaction.Source.USER);
+    log.info("Creating allocation transaction for fund '{}' and budget '{}' with allocation '{}'",
+      financeData.getFundId(), financeData.getBudgetId(), allocation);
+
+    return fiscalYearService.getFiscalYearById(financeData.getFiscalYearId(), requestContext)
```

Comment:
you can add validation to have the same  fiscalYearId for all objects in the payload, in this case it will not be needed to make requests for each entity

Reply:
Added vailidation and improved transaction creation